src/types.c

```diff
@@ -42,12 +45,55 @@ const char* vec_type_as_str(enum vctrs_type type) {
   case vctrs_type_raw:       return "raw";
   case vctrs_type_list:      return "list";
   case vctrs_type_dataframe: return "dataframe";
+  case vctrs_type_s3:        return "s3";
+  case vctrs_type_scalar:    return "scalar";
+  }
+}
+
+static SEXP vec_is_vector_dispatch_fn = NULL;
+
+bool vec_is_vector(SEXP x) {
+  switch (vec_typeof(x)) {
+  case vctrs_type_null:
+  case vctrs_type_scalar:
+    return false;
+  case vctrs_type_logical:
+  case vctrs_type_integer:
+  case vctrs_type_double:
+  case vctrs_type_complex:
+  case vctrs_type_character:
+  case vctrs_type_raw:
+  case vctrs_type_list:
+  case vctrs_type_dataframe:
+    return true;
+  case vctrs_type_s3:
+    if (Rf_inherits(x, "vctrs_vctr")) {
```

Comment:
Should this be another case in the enum?

Reply:
I think we'll generally need to dispatch for `vctrs_vctr` objects so I kept it as S3 for now.

Reply:
Yeah, we'll need to think that through.